### PR TITLE
Ensure that the specified rustup toolchain exists before using it

### PR DESCRIPTION
**Pull request by Ma27**
Given a config like this

    [tasks.test]
    command = "cargo"
    args = [ "build" ]
    toolchain = "alarm"

you get a fairly unintuitive error output like this:

```
[cargo-make] INFO - Running Task: test
[cargo-make] ERROR - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.
```

This can be quite confusing especially for people who are new to a
project and might've missed installing the specified rustup toolchain
before building the project with `cargo-make`.

This patch ensures in the `toolchain`-subcrate that the toolchain
actually exists (by checking the exit-code of `rustup run <toolchain>
true`) before running the actual task. If the toolchain doesn't exist,
an error like this will be thrown:

```
[cargo-make] INFO - Running Task: test
[cargo-make] ERROR - Please install the toolchain alarm using rustup before building!
[cargo-make] WARN - Build Failed.
```

### CODE REVIEW

**Comment by sagiegurari** [comment]
that's awesome. thanks a lot.
I'll review the PR but overall i think its really good idea.

File: src/lib/toolchain.rs [hunk 1]
```diff
@@ -8,12 +8,17 @@
...
 use crate::types::CommandSpec;
+use std::process::{Command, Stdio};
 
 pub(crate) fn wrap_command(
     toolchain: &str,
     command: &str,
     args: &Option<Vec<String>>,
 ) -> CommandSpec {
+    if !has_toolchain(toolchain) {
+        error!("Please install the toolchain {} using rustup before building!", &toolchain);
```
Comment (sagiegurari) [inline src/lib/toolchain.rs]: maybe people are not building anything but using this as a simple task runner.
maybe rephrase to: "Error: missing toolchain: {}" which is really simple message.

File: src/lib/toolchain_test.rs [hunk 2]
```diff
@@ -1,46 +1,59 @@
 use super::*;
+use crate::test;
+
+#[test]
+#[should_panic]
+fn invalid_toolchain() {
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: lets rename this to
wrap_command_invalid_toolchain
this makes it much easier for me to understand which test failed based on name.

File: src/lib/toolchain_test.rs [hunk 3]
```diff
@@ -1,46 +1,59 @@
...
+    wrap_command("invalid-chain", "true", &None);
+}
 
 #[test]
 fn wrap_command_none_args() {
-    let output = wrap_command("mychain", "testcommand", &None);
+    if test::is_not_rust_stable() {
+        let output = wrap_command("nightly", "true", &None);
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: this breaks the build on beta (see travis/appveyor builds). lets change to envmnt::get_or_panic("CARGO_MAKE_RUST_CHANNEL")

File: src/lib/toolchain_test.rs [hunk 4]
```diff
@@ -1,46 +1,59 @@
...
+#[test]
+#[should_panic]
+fn invalid_toolchain() {
+    wrap_command("invalid-chain", "true", &None);
+}
 
 #[test]
 fn wrap_command_none_args() {
-    let output = wrap_command("mychain", "testcommand", &None);
+    if test::is_not_rust_stable() {
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: lets remove this and test on all rust versions.

File: src/lib/toolchain_test.rs [hunk 5]
```diff
@@ -1,46 +1,59 @@
...
+        assert_eq!(args[0], "run".to_string());
+        assert_eq!(args[1], "nightly".to_string());
+        assert_eq!(args[2], "true".to_string());
+    }
 }
 
 #[test]
 fn wrap_command_empty_args() {
-    let output = wrap_command("mychain", "testcommand", &Some(vec![]));
+    if test::is_not_rust_stable() {
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: same. lets remove the if

File: src/lib/toolchain_test.rs [hunk 6]
```diff
@@ -1,46 +1,59 @@
...
+        let output = wrap_command("nightly", "true", &Some(vec![]));
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: same, lets use CARGO_MAKE_RUST_CHANNEL

File: src/lib/toolchain_test.rs [hunk 7]
```diff
@@ -1,46 +1,59 @@
...
-    assert_eq!(output.command, "rustup".to_string());
-
-    let args = output.args.unwrap();
-    assert_eq!(args.len(), 5);
-    assert_eq!(args[0], "run".to_string());
-    assert_eq!(args[1], "mychain".to_string());
-    assert_eq!(args[2], "testcommand".to_string());
-    assert_eq!(args[3], "echo".to_string());
-    assert_eq!(args[4], "test".to_string());
+    if test::is_not_rust_stable() {
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: same

File: src/lib/toolchain_test.rs [hunk 8]
```diff
@@ -1,46 +1,59 @@
...
+        let output = wrap_command(
+            "nightly",
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: same

File: src/lib/toolchain_test.rs (hunk 3 above)
Comment (Ma27) [reply src/lib/toolchain_test.rs]: Thanks! This was the env var I wasn't aware of which is why I implemented this gross hack in the tests :)

File: src/lib/toolchain.rs (hunk 1 above)
Comment (Ma27) [reply src/lib/toolchain.rs]: Fixed :+1:

File: src/lib/toolchain_test.rs [hunk 9]
```diff
@@ -39,8 +49,8 @@ fn wrap_command_with_args() {
     let args = output.args.unwrap();
     assert_eq!(args.len(), 5);
     assert_eq!(args[0], "run".to_string());
-    assert_eq!(args[1], "mychain".to_string());
-    assert_eq!(args[2], "testcommand".to_string());
+    assert_eq!(args[1], "nightly".to_string());
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: this needs to change to channel for tests to pass

File: src/lib/toolchain_test.rs [hunk 10]
```diff
@@ -1,36 +1,46 @@
...
+    let output = wrap_command(&channel, "true", &Some(vec![]));
 
     assert_eq!(output.command, "rustup".to_string());
 
     let args = output.args.unwrap();
     assert_eq!(args.len(), 3);
     assert_eq!(args[0], "run".to_string());
-    assert_eq!(args[1], "mychain".to_string());
-    assert_eq!(args[2], "testcommand".to_string());
+    assert_eq!(args[1], "nightly".to_string());
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: this needs to change to channel for tests to pass

File: src/lib/toolchain_test.rs [hunk 11]
```diff
@@ -1,36 +1,46 @@
...
+    let output = wrap_command(&channel, "true", &None);
 
     assert_eq!(output.command, "rustup".to_string());
 
     let args = output.args.unwrap();
     assert_eq!(args.len(), 3);
     assert_eq!(args[0], "run".to_string());
-    assert_eq!(args[1], "mychain".to_string());
-    assert_eq!(args[2], "testcommand".to_string());
+    assert_eq!(args[1], "nightly".to_string());
```
Comment (sagiegurari) [inline src/lib/toolchain_test.rs]: this needs to change to channel for tests to pass

File: src/lib/toolchain_test.rs (hunk 9 above)
Comment (Ma27) [reply src/lib/toolchain_test.rs]: Oof, sorry for that, just pushed a fix :)

**Comment by sagiegurari** [comment]
thanks a lot for the PR. i'm merging it now. don't worry about the build errors, i'll resolve them.
